source/simulator/lateral_averaging.cc

```diff
@@ -448,17 +250,279 @@ namespace aspect
         std::vector<Tensor<1,dim> > temperature_gradients;
         std::vector<double> temperature_values;
     };
+  }
+
+  template <int dim>
+  void LateralAveraging<dim>::compute_lateral_averages(std::vector<std_cxx11::unique_ptr<FunctorBase<dim> > > &functors,
+                                                       std::vector<std::vector<double> > &values) const
+  {
+    Assert (values.size() > 0,
+            ExcMessage ("To call this function, you need to request a positive "
+                        "number of properties to compute."));
+    Assert (values[0].size() > 0,
+            ExcMessage ("To call this function, you need to request a positive "
+                        "number of depth slices."));
+    Assert (functors.size() == values.size(),
+            ExcMessage ("To call this function, you need supply as many data "
+                        "vectors as properties you want to compute."));
+
+    const unsigned int n_properties = values.size();
+    const unsigned int n_slices = values[0].size();
+    for (unsigned int i=0; i < n_properties; ++i)
+      Assert (values[i].size() == n_slices,
+              ExcMessage ("All properties need to have the same number of depth slices."));
+
+    std::vector<double> volume(n_slices);
+
+    // this yields 10^dim quadrature points evenly distributed in the interior of the cell.
+    // We avoid points on the faces, as they would be counted more than once.
+    const QIterated<dim> quadrature_formula (QMidpoint<1>(),
+                                             10);
```

Comment:
Is this what we had before? Ten seems kind of a lot. If it's still slow, there would be an easy way to reduce the computational complexity.

Reply:
Yes, this is what I meant in my initial comment. I thought about adaptively computing the number of quadrature points based on the largest cell diameter and the size of the depth slices to make sure we use the lowest possible resolution while still having at least one quadrature point per depth slice. Unfortunately this decreases the accuracy of the postprocessor (because the volume fraction that we add to the wrong depth slice increases). @jdannberg preference was to have a slower but more accurate postprocessor. I am not sure about this, we could also make that an input argument.

Reply:
Yes, that's reasonable. After your 10x speedup, does the depth averaging still show up in your profiles?

Reply:
It is still 3 times as expensive as visualization, in my example model that is 11% of total runtime, if you run it every timestep. However, I suspect this is not longer crucial as most people will run it at most every 20 timesteps or so. In other words not terribly concerning right now. I thought about better ways to do this, but could not come up with one. In theory a `QAnisotropic` would save a lot of time, since we only need a high resolution in depth direction, but I think we can not rely on the orientation of the cells. Anyway, maybe I just stop here and say it is good enough.

Reply:
Yes, good enough. If someone complains, then we'll drop the number of integration points to 8 in each direction and gain a factor of 2 :-)